include/seqan3/alphabet/concept.hpp

```diff
@@ -14,24 +14,668 @@
+    //!\cond
+        -> decltype(impl(priority_tag<2>{}, s_alph_t{}, dummy{}))
+    //!\endcond
+    {
+        static_assert(noexcept(impl(priority_tag<2>{}, s_alph_t{})),
+            "Only overloads that are marked noexcept are picked up by seqan3::alphabet_size.");
+        static_assert(std::Constructible<size_t, decltype(impl(priority_tag<2>{}, s_alph_t{}))>,
+            "The return type of your alphabet_size implementation must be convertible to size_t.");
+        static_assert(SEQAN3_IS_CONSTEXPR(impl(priority_tag<2>{}, s_alph_t{})),
+            "Only overloads that are marked constexpr are picked up by seqan3::alphabet_size.");
+
+        return impl(priority_tag<2>{}, s_alph_t{});
+    }
+};
+
+} // namespace seqan3::detail::adl::only
+
+namespace seqan3
+{
+
+/*!\brief A type trait that holds the size of a (semi-)alphabet.
+ * \tparam your_type The (semi-)alphabet type being queried.
+ * \ingroup alphabet
+ *
+ * \details
+ *
+ * This type trait is implemented as a global variable template.
+ *
+ * It is only defined for types that provide one of the following (checked in this order):
+ *
+ *   1. A free function `alphabet_size(your_type const &)` in the namespace of your type (or as `friend`) that
+ *      returns the size as an integral value. The function must be marked `constexpr` and `noexcept` and the return
+ *      type needs to be implicitly convertible to `size_t`. The value of the argument to the function shall be ignored,
+ *      it is only used to select the function via
+ *      [argument-dependent lookup](https://en.cppreference.com/w/cpp/language/adl).
+ *   2. A free function `alphabet_size(your_type const &)` in `namespace seqan3::adaptation` that returns
+ *      the size as an integral value. Used only by the SeqAn library internally. The same restrictions apply as above.
+ *   3. A `static constexpr` data member called `alphabet_size` that is the size. It must
+ *      be implicitly convertible to `size_t`.
+ *
+ * Every (semi-)alphabet type must provide one of the above.
+ *
+ * *Note* that if the (semi-)alphabet type with cvref removed is not std::is_nothrow_default_constructible or not
+ * seqan3::is_constexpr_default_constructible, this object will instead look for
+ * `alphabet_size(std::type_identity<your_type> const &)` with the same semantics (in cases 1. and 2.).
+ *
+ * ### Example
+ *
+ * \include test/snippet/alphabet/alphabet_size.cpp
+ *
+ * For an example of a full alphabet definition with free function implementations (solution 1. above),
+ * see seqan3::Alphabet.
+ *
+ * ### Customisation point
+ *
+ * This is a customisation point (TODO link to manual). To specify the behaviour for your own alphabet type,
+ * simply provide one of the three functions specified above.
+ */
+template <typename alph_t>
+//!\cond
+    requires requires { { detail::adl::only::alphabet_size_fn<alph_t>{}() }; }
+//!\endcond
+inline constexpr auto alphabet_size_v = detail::adl::only::alphabet_size_fn<alph_t>{}();
```

Comment:
This does not cause trouble when calling the templatized operator() as non-template?

Reply:
I would have guessed, since you have a template operator that you need to call it with template in a template context.

```cpp
detail::adl::only::alphabet_size_fn<alph_t>{}.template operator()<>();
```
I mean it seems to be valid like it is. Just not sure if all compiler would allow this, but we will figure it out at some point I guess 😄

Reply:
It works when the types are deducted from arguments so I assumed defaulted template parameters would work as well...